api/grpcserver/mesh_service.go

```diff
@@ -610,3 +614,68 @@ func (s MeshService) EpochStream(req *pb.EpochStreamRequest, stream pb.MeshServi
 	)
 	return nil
 }
+
+func (s MeshService) MalfeasanceQuery(ctx context.Context, req *pb.MalfeasanceRequest) (*pb.MalfeasanceResponse, error) {
+	var id types.NodeID
+	if len(req.SmesherHex) == 0 {
+		if len(req.Smesher.Id) != types.NodeIDSize {
+			return nil, status.Error(codes.InvalidArgument,
+				fmt.Sprintf("invalid smesher id length (%d), expected (%d)", len(req.Smesher.Id), types.NodeIDSize))
+		}
+		id = types.BytesToNodeID(req.Smesher.Id)
+	} else {
+		parsed, err := hex.DecodeString(req.SmesherHex)
+		if err != nil {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+		id = types.BytesToNodeID(parsed)
+	}
+	proof, err := s.cdb.GetMalfeasanceProof(id)
+	if err != nil && !errors.Is(err, sql.ErrNotFound) {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	return &pb.MalfeasanceResponse{
+		Proof: events.ToMalfeasancePB(id, proof, req.IncludeProof),
+	}, nil
+}
+
+func (s MeshService) MalfeasanceStream(req *pb.MalfeasanceStreamRequest, stream pb.MeshService_MalfeasanceStreamServer) error {
+	sub := events.SubscribeMalfeasance()
+	if sub == nil {
+		return status.Errorf(codes.FailedPrecondition, "event reporting is not enabled")
+	}
+	eventch, fullch := consumeEvents[events.EventMalfeasance](stream.Context(), sub)
```

Comment:
NIT (and unrelated to this PR):

I don't think it makes sense to have `eventch` and `fullch` as return value in `consumeEvents` if the buffer in `eventch` runs full, the listener doesn't read from it. Why should the listener read from `fullch`?

The buffer is also huge -> 32768 events. So if `fullch` is closed the listener hasn't read events for quite some time.

Reply:
i see this more as a safeguard to stopgap a programming error from OOM e.g. doing something blocking when processing an event. not sure about the history here.

Reply:
Yeah, I am not sure about the current event handling. I see 2 possible approaches:

- Events can be dropped / there is no guarantee that every event will be sent to every consumer. This is the "safe" approach since busy consumers won't cause the producer to hang because they are not listening
- Events cannot be dropped / every event is guaranteed to be sent to every consumer. This way a malfunctioning consumer might cause the producer to get stuck because it cannot send more events.

Both approaches can give some leeway to busy consumers by using buffered channels. Here it looks like the first approach was taken (if the consumer isn't listening eventually the producer will stop sending). Just that the consumer will also not handle the events any more that were buffered for it when it becomes ready again. Additionally his happens not immediately: the channels are read in the same select statement, so some events might still be processed before the consumer gets the message that no events will be coming any more and stops reading from `eventch`.

Reply:
other stream APIs are implemented this way as well. if you feel strongly about changing the behavior, maybe file an issue for what you want to fix?

Reply:
the idea is that stream subscriber should never be able to block consensus code and we shouldn't silently drop events either. so once we see that subscriber is not reading fast enough error is returned from api and channel is unregistered